src/libYARP_sig/include/yarp/sig/PointCloud.h

```diff
@@ -0,0 +1,590 @@
+/*
+ * Copyright (C) 2006-2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+
+#ifndef YARP_SIG_POINTCLOUD_H
+#define YARP_SIG_POINTCLOUD_H
+
+#include <yarp/sig/Vector.h>
+#include <yarp/sig/PointCloudNetworkHeader.h>
+
+#include <map>
+#include <vector>
+#include <cstring>
+
+
+namespace yarp {
+    namespace sig {
+        template <class T> class PointCloud;
+        class PointCloudBase;
+    }
+}
+
+namespace yarp {
+    namespace sig {
+    /**
+     * @brief The PointCloudBase class
+     *
+     * A Base class for a yarp::sig::PointCloud<T>, provide default implementation for
+     * some methods and hides some internal implementation to the user.
+     * It can't be instantiated, only inherited.
+     */
+    class YARP_sig_API PointCloudBase : public yarp::os::Portable
+    {
+    public:
+
+        virtual size_t wireSizeBytes() const = 0;
+
+        virtual size_t dataSizeBytes() const = 0;
+
+        virtual size_t size() const = 0;
+
+        virtual const char* getRawData() const = 0;
+
+        virtual bool read(yarp::os::ConnectionReader& connection) override = 0;
+
+        virtual bool write(yarp::os::ConnectionWriter& writer) override = 0;
+
+        virtual int getBottleTag() const = 0;
+
+        /**
+         * @return the height specified in the yarp::sig::PointCloudNetworkHeader.
+         */
+        virtual size_t height() const;
+
+        /**
+         * @return the width specified int the yarp::sig::PointCloudNetworkHeader.
+         */
+        virtual size_t width() const;
+
+        /**
+         * @return the enum representing the point type.
+         */
+        virtual int getPointType() const;
+
+        virtual yarp::os::Type getType() override;
+
+        /**
+         * @return true if the point cloud is organized in an image-like structure
+         */
+        virtual bool isOrganized();
+
+        /**
+         * @return true if the point cloud doesn't contain NaN or Inf values
+         */
+        virtual inline bool isDense() const
+        {
+            return header.isDense;
+        }
+
+    protected:
+
+        PointCloudBase();
+
+        virtual void copyFromRawData(const char* dst, const char* source, std::vector<int> &recipe);
+
+        virtual std::vector<int> getComposition(int type_composite) const;
+
+        virtual size_t pointType2Size(int type) const;
+
+        virtual size_t getOffset(int type_composite, int type_basic) const;
+
+        yarp::sig::PointCloudNetworkHeader    header;
```

Comment:
This member should be hidden using pimpl, and perhaps accessed by subclasses using some method